### PR TITLE
dev: address address-related todos

### PR DESCRIPTION
**Pull request by enitrat**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #571 

## What is the new behavior?

- Addressed pending TODOs related to address calculation
- fixed some tests
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo [hunk 1]
```diff
@@ -1079,3 +1078,26 @@ fn test_exec_extcodehash_ca_with_bytecode() {
         'expected counter SC code hash'
     );
 }
+
+#[test]
+#[available_gas(2000000000)]
+fn test_exex_extcodehash_precompiles() {
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_environment_information.cairo:1084]: typo

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo [hunk 2]
```diff
@@ -1079,3 +1078,26 @@ fn test_exec_extcodehash_ca_with_bytecode() {
...
+    loop {
+        if i == 0x10 {
+            break;
+        }
+        machine.stack.push(i.into());
+        // When
+        machine.exec_extcodehash().unwrap();
+
+        // Then
+        assert(machine.stack.pop().unwrap() == 0, 'expected counter SC code hash');
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_environment_information.cairo:1100]: wrong error msg

File: crates/evm/src/tests/test_instructions/test_system_operations.cairo [hunk 3]
```diff
@@ -598,15 +600,38 @@ fn test_exec_selfdestruct_existing_ca() {
...
-// for now we can't fund an undeployed CA as the SN address is not deterministically calculated in the runner, so no way of funding a SN address by calculating it from an EVM address
-// This test should
-// - deploy kkt and token, deploy an EOA
-// - call `get_account` on an undeployed account, set its type to CA, its nonce to 1, its code to something
-// - selfdestruct it
-// verify that the value transfer has succeeded
-//
+fn test_selfdestruct_undeployed_ca() {
+    let (native_token, kakarot_core) = setup_contracts_for_testing();
+    let evm_address: EthAddress = 'ca_address'.try_into().unwrap();
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_system_operations.cairo:605]: feels a bit bad to write them by hand in every single test, i vote for refactoring them in some util getter

File: crates/evm/src/tests/test_instructions/test_system_operations.cairo [hunk 4]
```diff
@@ -598,15 +600,38 @@ fn test_exec_selfdestruct_existing_ca() {
...
+    let ca_address: Address = Address {
+        evm: evm_address, starknet: kakarot_core.compute_starknet_address(evm_address)
+    };
+    let recipient_address: EthAddress = 'recipient_address'.try_into().unwrap();
+    let recipient = deploy_eoa(recipient_address);
+    let ca_balance = 1000;
+    fund_account_with_native_token(ca_address.starknet, native_token, ca_balance);
+    let mut machine = MachineBuilderTestTrait::new_with_presets().with_target(ca_address).build();
+    // - call `get_account` on an undeployed account, set its type to CA, its nonce to 1, its code to something
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_system_operations.cairo:614]: its nonce to 0?

Comment (Eikix) [reply crates/evm/src/tests/test_instructions/test_system_operations.cairo:614]: its code to an empty bytes span

File: crates/evm/src/tests/test_model.cairo [hunk 5]
```diff
@@ -119,26 +119,26 @@ fn test_account_exists_contract_account() {
...
+fn test_account_has_code_or_nonce_undeployed() {
     // Given
     let (native_token, kakarot_core) = setup_contracts_for_testing();
 
     // When
     let account = AccountTrait::fetch_or_create(evm_address());
 
     // Then
-    assert(account.exists() == false, 'account should be deployed');
+    assert(account.has_code_or_nonce() == false, 'account should be deployed');
```
Comment (Eikix) [inline crates/evm/src/tests/test_model.cairo:136]: false error msg

File: crates/evm/src/tests/test_model.cairo [hunk 6]
```diff
@@ -119,26 +119,26 @@ fn test_account_exists_contract_account() {
     let account = AccountTrait::fetch(evm_address()).unwrap();
 
     // Then
-    assert(account.exists() == true, 'account should be deployed');
+    assert(account.has_code_or_nonce() == true, 'account should be deployed');
```
Comment (Eikix) [inline crates/evm/src/tests/test_model.cairo:122]: inexact error message

**Review by Eikix — COMMENTED** [review]
a few typos, but overall gg lgtm

File: crates/evm/src/tests/test_model.cairo (hunk 6 above)
Comment (enitrat) [reply crates/evm/src/tests/test_model.cairo:122]: well not really, it's actually accurate for that matter

Comment (Eikix) [reply crates/evm/src/tests/test_model.cairo:122]: ah my bad thought CA would be in state local storage before it is deployed

Comment (enitrat) [reply crates/evm/src/tests/test_model.cairo:122]: hm yes ok maybe its not _precise_ enough but deployed in the sense, kakarot-deployed

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo (hunk 1 above)
Comment (enitrat) [reply crates/evm/src/tests/test_instructions/test_environment_information.cairo:1084]: fixed

File: crates/evm/src/tests/test_instructions/test_system_operations.cairo (hunk 3 above)
Comment (enitrat) [reply crates/evm/src/tests/test_instructions/test_system_operations.cairo:605]: you mean the contract addresses or the overall setup logic?

File: crates/evm/src/tests/test_model.cairo (hunk 6 above)
Comment (enitrat) [reply crates/evm/src/tests/test_model.cairo:122]: done

**Review by Eikix — APPROVED** [review]
lgtm

File: crates/evm/src/tests/test_instructions/test_system_operations.cairo (hunk 3 above)
Comment (Eikix) [reply crates/evm/src/tests/test_instructions/test_system_operations.cairo:605]: the contract addresses

**Pull request merged**